experimental/iterators/lib/Conversion/IteratorsToStandard/IteratorsToStandard.cpp

```diff
@@ -136,6 +191,104 @@ struct IteratorConversionPattern : public ConversionPattern {
   StringRef destructorName;
 };
 
+struct ConstantTupleLowering : public ConversionPattern {
+  ConstantTupleLowering(MLIRContext *context, PatternBenefit benefit = 1)
+      : ConversionPattern("iterators.constant", benefit, context) {}
+
+  LogicalResult
+  matchAndRewrite(Operation *op, ArrayRef<Value> operands,
+                  ConversionPatternRewriter &rewriter) const override {
+    // Create LLVM struct type.
+    assert(op->getNumResults() == 1);
+    TupleType resultType = op->getResult(0).getType().dyn_cast<TupleType>();
+    assert(resultType);
+    LLVMStructType structType = LLVMStructType::getNewIdentified(
+        getContext(), "tuple", resultType.getTypes());
+
+    // Undef.
+    Value structValue = rewriter.create<UndefOp>(op->getLoc(), structType);
+
+    // Insert values.
+    ArrayAttr values = op->getAttr("values").dyn_cast<ArrayAttr>();
+    assert(values);
+    for (int i = 0; i < static_cast<int>(values.size()); i++) {
+      // Create index attribute.
+      ArrayAttr indicesAttr = rewriter.getIndexArrayAttr({i});
+
+      // Create constant value op.
+      Value valueOp = rewriter.create<LLVM::ConstantOp>(
+          op->getLoc(), values[i].getType(), values[i]);
+
+      // Insert into struct.
+      structValue = rewriter.create<InsertValueOp>(op->getLoc(), structValue,
+                                                   valueOp, indicesAttr);
+    }
+
+    rewriter.replaceOp(op, structValue);
+
+    return success();
+  }
+};
+
+struct PrintOpLowering : public ConversionPattern {
+  PrintOpLowering(MLIRContext *context, PatternBenefit benefit = 1)
+      : ConversionPattern("iterators.print", benefit, context) {}
+
+  LogicalResult
+  matchAndRewrite(Operation *op, ArrayRef<Value> operands,
+                  ConversionPatternRewriter &rewriter) const override {
+    assert(op->getNumOperands() == 1);
+    TupleType tupleType = op->getOperand(0).getType().dyn_cast<TupleType>();
+    assert(tupleType);
+
+    assert(operands.size() == 1);
+    LLVMStructType structType =
+        operands[0].getType().dyn_cast<LLVMStructType>();
+    assert(structType);
+
+    // Assemble format string in the form `(%i, %i, ...)`.
+    SmallString<8> format("(");
```

Comment:
you could use llvm::interleaveComma on a stringstream here to handle the comma logic in a more idiomatic fashion